service/topology_coordinator.cc

```diff
@@ -2044,6 +2060,100 @@ future<bool> topology_coordinator::maybe_start_tablet_migration(group0_guard gua
     co_return true;
 }
 
+future<locator::load_stats> topology_coordinator::refresh_tablet_load_stats() {
+    auto tm = get_token_metadata_ptr();
+    auto& topology = tm->get_topology();
+
+    locator::load_stats stats;
+    static constexpr std::chrono::seconds wait_for_live_nodes_timeout{30};
+
+    for (auto& [dc, nodes] : topology.get_datacenter_nodes()) {
+        locator::load_stats dc_stats;
+        rtlogger.info("raft topology: Refreshing table load stats for DC {} that has {} endpoints", dc, nodes.size());
+        co_await coroutine::parallel_for_each(nodes, [&] (const auto& node) -> future<> {
+            auto dst = node->host_id();
+
+            _as.check();
+
+            // FIXME: if a node is down, completion status cannot be relied upon, but the average tablet size
+            //  could still be inferred from the replicas available.
+
+            auto timeout = netw::messaging_service::clock_type::now() + wait_for_live_nodes_timeout;
+
+            abort_source as;
+            auto request_abort = [&as] () mutable noexcept {
+                as.request_abort();
+            };
+            auto t = timer<lowres_clock>(request_abort);
+            t.arm(timeout);
+            auto sub = _as.subscribe(request_abort);
+
+            auto node_stats = co_await ser::storage_service_rpc_verbs::send_table_load_stats(&_messaging,
+                                                                                             netw::msg_addr(id2ip(dst)),
+                                                                                             as,
+                                                                                             raft::server_id(dst.uuid()));
+
+            dc_stats += node_stats;
+        });
+
+        for (auto& [table_id, table_stats] : dc_stats.tables) {
+            co_await coroutine::maybe_yield();
+
+            auto& t = _db.find_column_family(table_id);
+            auto& rs = t.get_effective_replication_map()->get_replication_strategy();
+            if (!rs.uses_tablets()) {
+                continue;
+            }
+            const auto* nts_ptr = dynamic_cast<const locator::network_topology_strategy*>(&rs);
+            if (!nts_ptr) {
+                on_internal_error(rtlogger, "Cannot convert replication_strategy that uses tablets into network_topology_strategy");
+            }
+
+            auto rf_for_this_dc = nts_ptr->get_replication_factor(dc);
+            if (rf_for_this_dc <= 0) {
+                continue;
+            }
+            // Takes into account the RF of each DC, so stats will still make sense when aggregated
+            // from multiple DCs. This allows tablet allocator to compute an average tablet size
+            // by dividing total size in bytes by tablet count.
+            table_stats.size_in_bytes /= rf_for_this_dc;
```

Comment:
Isn't it sufficient to divide by total rf in the end?

AFAICT the stats you return have a factor of dcs_that_replicate_this_keyspace, is it adjusted later?

Better to normalize not just within the datacenter, but overall.

Reply:
Assume we have two DCs, DC1 with RF=2, and DC2 with RF=3

Tablet count for table T is 8

let's say, for simplicy, all tablets are 1G in size

Then DC1 has 8*RF(2)=16G for table T

And DC2 has  8*RF(3)=24g for table T

coordinator retrieves 16G in DC1 and divide it by 2, 8G

""          retrieves 24G in DC2 and divide it by 3, same 8G

now we have to divide it by dc count, which is the part missing. thanks!

alternatively, I divide (total size across DCs) by (total sum of replicas across DCs).

Reply:
Yes that's what I meant about normalize overall (divide by global RF).